src/dune/file_tree.ml

```diff
@@ -327,104 +295,249 @@ let make_root
+    | None -> vcs
+    | Some kind -> Some { Vcs.kind; root = Path.(append_source root) path }
+
+  let make_subdir =
+    let module Input = struct
+      type t = Output.t * Sub_dirs.Status.t * string
+
+      let hash : t -> int =
+       fun ((dir, _), _, s) ->
+        Tuple.T2.hash Path.Source.hash String.hash (dir.path, s)
+
+      let to_dyn (dir, status, name) =
+        let open Dyn.Encoder in
+        record
+          [ ("dir", Output.to_dyn dir)
+          ; ("status", Sub_dirs.Status.to_dyn status)
+          ; ("name", String.to_dyn name)
+          ]
+
+      let equal (((d1, _), _, n1) : t) (((d2, _), _, n2) : t) =
+        Path.Source.equal d1.path d2.path && String.equal n1 n2
+    end in
+    let f
+        ( ((parent_dir, dirs_visited) : Output.t)
+        , (dir_status : Sub_dirs.Status.t)
+        , name ) =
+      let path = Path.Source.relative parent_dir.path name in
+      let settings = Settings.get () in
+      let readdir =
+        match Readdir.of_source_path path with
+        | Ok dir -> dir
+        | Error _ -> Readdir.empty
+      in
+      let project =
+        if dir_status = Data_only then
+          parent_dir.project
+        else
+          Option.value
+            (Dune_project.load ~dir:path ~files:readdir.files
+               ~infer_from_opam_files:settings.recognize_jbuilder_projects)
+            ~default:parent_dir.project
+      in
+      let vcs = get_vcs ~default:parent_dir.vcs ~readdir ~path in
+      let contents, visited =
+        contents readdir ~dirs_visited ~project ~path ~dir_status
+      in
+      let dir = Dir0.create ~project ~path ~status:dir_status ~contents ~vcs in
+      (dir, visited)
     in
-    walk path
-      ~dirs_visited:(File.Map.singleton (File.of_source_path path) path)
-      ~dir_status:Normal ~project ~vcs:ancestor_vcs x
-  in
-  match walk with
-  | Ok dir -> dir
-  | Error m ->
-    User_error.raise
-      [ Pp.textf "Unable to load source %s.@.Reason:%s@."
-          (Path.Source.to_string_maybe_quoted path)
-          (Unix.error_message m)
-      ]
-
-let get =
-  let memo =
-    Memo.create "file-tree" ~doc:"file tree"
-      ~input:(module Unit)
-      ~visibility:Memo.Visibility.Hidden
-      ~output:(Simple (module Dir))
-      Sync
-      (fun () ->
-        let (_ : Memo.Run.t) = Memo.current_run () in
-        make_root (Fdecl.get Settings.t))
-  in
-  Memo.exec memo
+    let memo =
+      Memo.create "make_subdir" ~doc:"make_subdir"
+        ~input:(module Input)
+        ~output:(Simple (module Output))
+        ~visibility:Memo.Visibility.Hidden Sync f
+    in
+    fun parent dir_status name -> Memo.exec memo (parent, dir_status, name)
+
+  let rec find_dir (t : Output.t) = function
+    | [] -> Some t
+    | comp :: components ->
+      let open Option.O in
+      let parent_dir, _ = t in
+      let* subdir = String.Map.find (Dir0.sub_dirs parent_dir) comp in
+      let subdir = make_subdir t subdir comp in
+      find_dir subdir components
```

Comment:
The component-by-component lookup might not be worth it anymore. In particular, we know do one memoized call per-component, which is likely to be higher than the cost of doing a single memoized call on the whole path.

Reply:
One thing that could be useful though is to allow to specify a custom dictionary data-structure in `Memo`, so that for some memoized function we use something better than a hash-table.

Reply:
> The component-by-component lookup might not be worth it anymore. In particular, we know do one memoized call per-component, which is likely to be higher than the cost of doing a single memoized call on the whole path.

I tried doing it this way initially, but it turned that I had to memoize this output type: `(Dir.t * visited) option`. To me this is a bit worse as it seems like we're polluting the cache with many meaningless dirs. Do you see another way to do it, or do you think memoizing the option type is fine?

Reply:
> One thing that could be useful though is to allow to specify a custom dictionary data-structure in Memo, so that for some memoized function we use something better than a hash-table.

Indeed. I'll make a ticket for this task

Reply:
The way I had imagined it would work is:

We have two memoised functions: `get_dir` and `get_dir_raw`. The first one is the user visible one and the second one the internal one. `get_dir` allows cutoffs but not `get_dir_raw`. `get_dir_raw` returns `(Dir.t * visited) option` and `git_dir` simply drops the `visited` part. If `p` is not the root, then `get_dir_raw p` ends up calling `get_dir_raw <parent of p>`.

Reply:
This way, to retrieve a directory we only need a single memoized call

Reply:
well, except the very first time, but for subsequent one we'll have a single call

Reply:
Okay, I think this is how I assumed it would work as well. A concrete disadvantage of this approach is that it now memoizes paths that don't exist. So for example, any kind of `File_tree.dir_exists` would now add an entry to the hash table. Do you think this cost is acceptable?